topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -133,72 +134,136 @@ public void init(Vertx vertx, Context context) {
      * Stop the operator.
      */
     @Override
-    @SuppressWarnings("deprecation") // Uses a deprecated executeBlocking call that should be addressed later. This is tracked in https://github.com/strimzi/strimzi-kafka-operator/issues/9233
     public void stop(Promise<Void> stop) throws Exception {
+        LOGGER.info("Stopping");
         this.stopped = true;
-        Long timerId = this.timerId;
-        if (timerId != null) {
-            vertx.cancelTimer(timerId);
+        if (this.timerId != null) {
+            vertx.cancelTimer(this.timerId);
         }
-        vertx.executeBlocking(blockingResult -> {
-            long timeout = 120_000L;
-            long deadline = System.currentTimeMillis() + timeout;
-            LOGGER.info("Stopping");
-            LOGGER.debug("Stopping kube watch");
-            topicWatch.close();
-            LOGGER.debug("Stopping zk watches");
-            topicsWatcher.stop();
-
-            Promise<Void> promise = Promise.promise();
-            Handler<Long> longHandler = new Handler<>() {
-                @Override
-                public void handle(Long inflightTimerId) {
-                    if (!topicOperator.isWorkInflight()) {
-                        LOGGER.debug("Inflight work has finished");
-                        promise.complete();
-                    } else if (System.currentTimeMillis() > deadline) {
-                        LOGGER.error("Timeout waiting for inflight work to finish");
-                        promise.complete();
-                    } else {
-                        LOGGER.debug("Waiting for inflight work to finish");
-                        vertx.setTimer(1_000, this);
-                    }
+        closeKubernetesWatch()
```

Comment:
My understanding of how the vertx futures work is that this future failing will result in all the subsequent `compose` methods not being called and every `onFailure` method being invoked. Is that the intended behaviour here?

Reply:
Yeah, this is also my understanding. Here we want to stop the internal components in a specific order and that's why we use compose. Each method of this chain completes successfully, regardless of the operation outcome. One could ask why we log failures then. I though that they may capture some framework level error. Does it make sense?

Reply:
Doesn't `disconnectFromZk` have the possibility to fail?

Reply:
I guess the other question is what we want to happen if a framework level error happens, in that case is it ok to skip stopping the remaining components? Or would we expect to attempt to stop the other components even if an earlier one fails?

Reply:
> Doesn't disconnectFromZk have the possibility to fail?

Right. That needs to be changed. Thanks.

> Would we expect to attempt to stop the other components even if an earlier one fails

I would say no. All external systems should be able to cope with unclean disconnections due to application failures or network issues, so I see this as a best effort attempt to orderly stop all components before application shutdown, with a configurable hard timeout (10 seconds by default).

Reply:
Thanks, that all makes sense to me